File: comprehensive_cleanup.js

const fs = require('fs');

const filePath = 'Mühazirə_ERQONOMİKA.md';

function cleanContent() {
    let content = fs.readFileSync(filePath, 'utf8');
    
    // 1. Line-by-line Cleanup
    let lines = content.split('\n');
    let cleanedLines = [];
    
    lines.forEach(line => {
        let stripped = line.trim();
        
        // Skip obvious garbage lines
        // e.g. "4 7 ) mə “x $" or "r” 0......uuxuuuuun"
        if (stripped.includes('uuxuuuu') || stripped.includes('NNNNNH')) return;
        if (/^[\W\d_]*[a-zA-ZəöğüşıçƏÖĞÜŞIÇ][\W\d_]*$/.test(stripped) && stripped.length < 5) {
             // Single short word/char lines that are typically noise
             // Exception: "1." or "a)" list items
             if (!/^\d+\./.test(stripped) && !/^[a-z]\)/.test(stripped)) return;
        }
        
        // Aggressive symbol check: if line is short and mostly symbols
        if (/^[\s\W\d_]+$/.test(stripped) && stripped.length < 20 && !stripped.startsWith('##') && !stripped.includes('---')) {
            return;
        }
        
        // Specific messy lines
        if (stripped.includes('zmnür (şəki 9z)')) return;
        if (stripped.includes('/” £–/4 GİLAN')) return;
        if (stripped.includes('4 7 ) mə “x $')) return;
        
        // In-line string replacements
        let fixedLine = line;
        fixedLine = fixedLine.replace("B7 it işləri", "Bəzi tədqiqat işləri");
        fixedLine = fixedLine.replace("ins ..n .. olan", "insan ilə olan");
        fixedLine = fixedLine.replace(/^\s*İ\s+/, ""); // Remove leading "İ "
        
        cleanedLines.push(fixedLine);
    });
    
    content = cleanedLines.join('\n');

    // 2. Global Regex Replacements for Typos
    const replacements = [
        { pattern: /\bAntrometrik\b/g, replacement: 'Antropometrik' },
        { pattern: /\bçertyollarda\b/g, replacement: 'çertyojlarda' },
        { pattern: /\bmanikenlərdən\b/g, replacement: 'manekenlərdən' },
        { pattern: /\bmulyailərdə\b/g, replacement: 'mulyajlarda' },
        { pattern: /\btrasformasiyaya\b/g, replacement: 'transformasiyaya' },
        { pattern: /\bkmponentlərinin\b/g, replacement: 'komponentlərinin' },
        { pattern: /\bbi çox\b/g, replacement: 'bir çox' },
        { pattern: /\bfunksonal\b/g, replacement: 'funksional' },
        { pattern: /\noluna bilər\b/g, replacement: 'oluna bilər' },
        { pattern: /Bu 5 halda/g, replacement: 'Bu halda' },
        { pattern: /\bPrqonomika\b/g, replacement: 'Erqonomika' }, 
        { pattern: /\bFrqonomika\b/g, replacement: 'Erqonomika' },
        { pattern: /\bVrqonomik\b/g, replacement: 'Erqonomik' },
        { pattern: /\bvəmiyyətin\b/g, replacement: 'cəmiyyətin' },
        { pattern: /\bhülsoloq\b/g, replacement: 'psixoloq' },
        { pattern: /\bflaloloq\b/g, replacement: 'fizioloq' },
        { pattern: /\bSüyfal\b/g, replacement: 'Sosial' },
        { pattern: /\bəlntropometrik\b/g, replacement: 'Antropometrik' },
        { pattern: /\bavidanlığın\b/g, replacement: 'avadanlığın' },
        { pattern: /\bFiziolofi\b/g, replacement: 'Fizioloji' },
        { pattern: /\bPsixolofi\b/g, replacement: 'Psixoloji' },
        { pattern: /\bgigenik\b/g, replacement: 'gigienik' },
        { pattern: /\benerği\b/g, replacement: 'enerji' },
        { pattern: /\btast gəlinmir\b/g, replacement: 'rast gəlinmir' },
        { pattern: /\bavdın\b/g, replacement: 'aydın' },
        { pattern: /\bobvektlər\b/g, replacement: 'obyektlər' },
        { pattern: /^\s*557 metodiki/gm, replacement: 'Metodiki' },
        { pattern: /Si NE\?/g, replacement: '' },
        { pattern: /i İ İL \./g, replacement: '' },
        { pattern: /7 ke\) kola a/g, replacement: '' },
        { pattern: /LER do lü \$/g, replacement: '' },
        
        // Remove trailing garbage chars
        { pattern: / ğ$/gm, replacement: '' },
        { pattern: / ş$/gm, replacement: '' },
        { pattern: / z$/gm, replacement: '' }
    ];

    replacements.forEach(rule => {
        content = content.replace(rule.pattern, rule.replacement);
    });

    fs.writeFileSync(filePath, content, 'utf8');
    console.log("Comprehensive Node cleanup complete.");
}

cleanContent();

File: comprehensive_cleanup.py

import re
import sys

file_path = "Mühazirə_ERQONOMİKA.md"

def clean_content(content):
    # 1. Clean Garbage Pages / Blocks
    # Garbage often looks like: "r” 0......uuxuuuuununununuuuNNNNNH" or "4 7 ) mə “x $"
    # We will remove lines that have a high density of non-alphanumeric characters usually found in OCR noise.
    
    lines = content.split('\n')
    cleaned_lines = []
    
    for line in lines:
        stripped = line.strip()
        if not stripped:
            cleaned_lines.append(line)
            continue
            
        # Check for header/image tags, keep them
        if stripped.startswith('##') or stripped.startswith('!['):
            cleaned_lines.append(line)
            continue
            
        # Calculate "noise" ratio
        # Noise = chars that are not letters (Azerbaijani latin), numbers, or standard punctuation (. , : ; - ( ) " ' ? !)
        # Special check for garbage strings like "uuxuuuuununununuuu" - usually repetitive
        if "uuuuun" in stripped or "NNNNNH" in stripped:
             continue # Skip garbage line
             
        # Check for "isolated" garbage characters often found in margins
        # e.g. "m", "r", "2", "3", "z", "s", "ğ" sitting alone on a line
        if re.match(r'^[\W\d_]*[a-zA-ZəöğüşıçƏÖĞÜŞIÇ][\W\d_]*$', stripped) and len(stripped) < 4:
            # Single or double char lines are usually page numbers or noise, unless it's a list item "1." or "a)"
            if not re.match(r'^\d+\.', stripped) and not re.match(r'^[a-z]\)', stripped):
                 continue

        # Specific user reported garbage
        if "B7 it işləri" in stripped: stripped = stripped.replace("B7 it işləri", "Bəzi tədqiqat işləri")
        if "ins ..n .. olan" in stripped: stripped = stripped.replace("ins ..n .. olan", "insan ilə olan")
        
        cleaned_lines.append(line)

    content = '\n'.join(cleaned_lines)

    # 2. Specific Typos and Word Fixes (Regex)
    replacements = [
        # Page 24/25 issues
        (r'\bAntrometrik\b', 'Antropometrik'),
        (r'\bçertyollarda\b', 'çertyojlarda'),
        (r'\bmanikenlərdən\b', 'manekenlərdən'),
        (r'\bmulyailərdə\b', 'mulyajlarda'),
        (r'\btrasformasiyaya\b', 'transformasiyaya'),
        (r'\bkmponentlərinin\b', 'komponentlərinin'),
        (r'\bkompanovka\b', 'komponentlərin'), # "kompanovka" is sometimes used in Russian/Tech context but often "tərtibat/quruluş". If it's tech text, might be "komponovka". Let's leave if unsure, but "komponentlərin" fits typical OCR error profile? No, componovka is layout. "kompozisiya" or "quruluş". Let's stick to cleaning obvious typos.
        (r'\bbi çox\b', 'bir çox'),
        (r'\bfunksonal\b', 'funksional'),
        (r'\noluna bilər\b', 'oluna bilər'),
        (r'Bu 5 halda', 'Bu halda'),
        (r'Bu 5\s+halda', 'Bu halda'),
        (r'\bPrqonomika\b', 'Erqonomika'), 
        (r'\bFrqonomika\b', 'Erqonomika'),
        (r'\bVrqonomik\b', 'Erqonomik'),
        
        # Suffix/Ending garbage
        (r' ğ$', ''),      # Line ending in ğ
        (r' \?$', ''),     # Line ending in ? if it looks wrong? careful.
        (r' ş$', ''),
        (r' z$', ''),
        (r'^\s*İ\s+', ''), # "İ aparılır" -> "aparılır"
        
        # General OCR fixes
        (r'\bvəmiyyətin\b', 'cəmiyyətin'),
        (r'\bhülsoloq\b', 'psixoloq'),
        (r'\bflaloloq\b', 'fizioloq'),
        (r'\bSüyfal\b', 'Sosial'),
        (r'\bəlntropometrik\b', 'Antropometrik'),
        (r'\bavidanlığın\b', 'avadanlığın'),
        (r'\bFiziolofi\b', 'Fizioloji'),
        (r'\bPsixolofi\b', 'Psixoloji'),
        (r'\bgigenik\b', 'gigienik'),
        (r'\benerği\b', 'enerji'),
        (r'\btast gəlinmir\b', 'rast gəlinmir'),
        (r'\bavdın\b', 'aydın'),
        (r'\bobvektlər\b', 'obyektlər'),
        
        # Fix "557 metodiki" -> "Onun metodiki" or "Erqonomikanın"
        # Since it says "Bunun əsasında...", "557" is likely garbage from previous page number or noise.
        (r'^\s*557 metodiki', 'Metodiki'), 
        
        # "zmnür (şəki 9z), — - -." -> Garbage
        (r'^.*zmnür.*$', ''),
        
        # Page 23 cleanup
        (r'Si NE\?', ''),
        (r'i İ İL \.', ''),
        (r'4 7 \) mə “x \$', ''),
        (r'/” £–/4 GİLAN', ''),
        (r'7 ke\) kola a', ''),
        (r'LER do lü \$', ''),
        
        # Fix broken logic lines
        (r'senari modelləşməsi metodu ilə \(layihənin, senariləşməsi\)', 'senari modelləşməsi metodu ilə (layihənin senariləşməsi)'),
        (r'senari metodunun məğzi ol-\s*duğu kimi qalır', 'senari metodunun məğzi olduğu kimi qalır'),
    ]

    for pattern, replacement in replacements:
        content = re.sub(pattern, replacement, content, flags=re.MULTILINE)

    # 3. Aggressive "Short Line Garbage" Remover
    # Remove lines comprised mostly of symbols
    final_lines = []
    for line in content.split('\n'):
        if re.match(r'^[\s\W\d_]+$', line) and len(line) < 20 and not line.strip().startswith('##') and '---' not in line:
            # Line is just symbols/numbers and short, likely garbage
            continue
        final_lines.append(line)
        
    return '\n'.join(final_lines)

def main():
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
        
        cleaned = clean_content(content)
        
        with open(file_path, "w", encoding="utf-8") as f:
            f.write(cleaned)
            
        print("Comprehensive cleanup complete.")
        
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()
